### DISCUSSION TITLE
@ConfigurationProperties의 장점은 무엇인가?

### DISCUSSION THREAD

**Discussion by alstn113** [2024-woowacourse-study/level-interview#169] [Spring Core]
.

**Comment by alstn113** [comment]
`@ConfigurationProperties`는 스프링 프레임워크에서 외부 설정 파일(application.properties 또는 application.yml)에 정의된 설정을 자바 빈 클래스에 매핑하는 데 사용되는 애노테이션입니다. 스프링 애플리케이션에서 설정값을 객체 지향적으로 관리하고, 설정의 타입 안전성, 유효성 검사, 재사용성 등을 제공합니다.

### 1. **타입-세이프 설정 바인딩**

- **강력한 타입 지원**:
    - 설정값이 자바 객체의 필드로 바인딩되므로, 타입 안전성을 확보할 수 있습니다. 
    이를 통해 설정값을 읽고 사용할 때 타입 변환 오류를 방지할 수 있습니다.
- **복잡한 타입 지원**:
    - 복잡한 계층 구조의 설정값도 객체 계층 구조에 매핑할 수 있습니다.

### 2. **유효성 검사**

- **`@Validated`와 함께 사용**:
    - `@ConfigurationProperties` 클래스에 `@Validated` 애노테이션을 사용하면, 설정값에 대한 유효성 검사를 할 수 있습니다.
- **설정값 검증**:
    - 설정값이 잘못된 경우 애플리케이션이 시작되지 않도록 하여, 오류를 조기에 발견할 수 있습니다.

### 3. **재사용성 및 모듈화**

- **설정 모듈화**:
    - 설정을 여러 클래스로 나누어 관리함으로써 설정의 모듈화를 할 수 있습니다. 이는 코드의 가독성과 유지보수성을 향상시킵니다.
- **재사용 가능한 구성**:
    - 공통 설정을 하나의 `@ConfigurationProperties` 클래스에 모아 여러 곳에서 재사용할 수 있습니다.

### 4. **유연한 바인딩 옵션**

- **리스트 및 맵 지원**:
    - 단순한 값뿐만 아니라, 리스트(List), 맵(Map) 같은 복잡한 자료 구조도 설정값으로 바인딩할 수 있습니다.

**Comment by 3Juhwan** [comment]
`@ConfigurationProperties`는 스프링 프레임워크에서 애플리케이션 설정 값을 매핑하기 위해 사용되는 어노테이션입니다. `@ConfigurationProperties`으로 선언된 클래스는 빈으로 등록하여 Application Context가 관리할 수 있도록 합니다. 

**타입 안전성**

`@ConfigurationProperties`를 사용하면 설정 파일(application.properties 또는 application.yml)에 정의된 설정 값들을 Java POJO 클래스로 매핑할 수 있습니다. 이는 컴파일 타임에 타입 체크가 가능하다는 의미이며, 잘못된 타입의 설정 값이 사용될 때 컴파일 오류가 발생하여 더 안전한 코드를 작성할 수 있습니다.

**다양한 설정 구조 지원**

`@ConfigurationProperties`는 복잡한 설정 구조를 쉽게 매핑할 수 있습니다. 중첩된 객체나 컬렉션을 포함한 다양한 설정 구조를 간단하게 매핑할 수 있습니다. 

**유연한 바인딩**

스프링은 바인딩 속성에 대해 완화된 규칙을 사용합니다. 예시로 `hostName` 필드에 `hostname`, `HOST_NAME` 등의 설정값을 바인딩할 수 있습니다. 또한 Lists, Maps, 그리고 Classes과 같은 복잡한 자료구조를 바인딩할 수 있습니다. 

**속성 변환**

여러 유형의 속성을 해당 빈에 바인딩하기 위한 변환을 지원합니다. 예시로 `9ns`는 Duration으로, `2GB`는 DataSize로, 그리고 전용 Converter도 선언하여 사용할 수 있습니다. 

**테스트 용이성**

설정 클래스를 이용하여 설정 값들을 쉽게 주입하고 모킹할 수 있어 테스트 코드 작성이 용이합니다.
